include/seqan3/alignment/data_structure/scoring_scheme_base.hpp

```diff
@@ -0,0 +1,220 @@
+//
+// Copyright (c) 2006-2018, Knut Reinert & Freie Universitaet Berlin
+// Copyright (c) 2016-2018, Knut Reinert & MPI Molekulare Genetik
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Provides seqan3::scoring_scheme_base.
+ * \author Hannes Hauswedell <hannes.hauswedell AT fu-berlin.de>
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <range/v3/algorithm/copy.hpp>
+
+#include <seqan3/alphabet/concept.hpp>
+#include <seqan3/core/detail/strong_type.hpp>
+#include <seqan3/core/concept/cereal.hpp>
+
+#if SEQAN3_WITH_CEREAL
+#include <cereal/types/array.hpp>
+#endif // SEQAN3_WITH_CEREAL
+
+namespace seqan3
+{
+
+/*!\brief A strong type of underlying type `int8_t` that represents the score two matching characters.
+ * \ingroup data_structure
+ * \see scoring_scheme_base::set_scheme_simple
+ */
+struct match_score : detail::strong_type<int8_t, match_score, detail::strong_type_skill::convert>
+{
+     using detail::strong_type<int8_t, match_score, detail::strong_type_skill::convert>::strong_type;
+};
+
+/*!\brief A strong type of underlying type `int8_t` that represents the score two different characters.
+ * \ingroup data_structure
+ * \see scoring_scheme_base::set_scheme_simple
+ */
+struct mismatch_score : detail::strong_type<int8_t, mismatch_score, detail::strong_type_skill::convert>
```

Comment:
Maybe `mismatch_penalty` instead of `score` since a negative value is expected?

Reply:
No, we shouldn't mix the terms. Either we use "score" **OR** we "cost" or "penalty". But mixing these is confusing because it implies inverting the sign (a penalty of 3 is a score of -3). If we have both we need to add two separate functions (`set_score` and `set_penalty`).

Reply:
So we will also call the gap config `gap_open_score` and `gap_extend_score` ?